### PR TITLE
ACNCD_Custom_DataConnector_v2

### PR DESCRIPTION
**Pull request by chicduong**
Fixes #

## Proposed Changes

  - Proofpoint TAP
  - Qualys VM
  - VMware Carbon Black

### CODE REVIEW

File: Workbooks/WorkbooksMetadata.json [hunk 1]
```diff
@@ -739,5 +739,44 @@
     "templateRelativePath": "SysmonThreatHunting.json",
     "subtitle": "",
     "provider": "Edoardo Gerosa"
+  },
+  {
+    "workbookKey": "VMwareCarbonBlackWorkbook",
+    "logoFileName": "vmwarecarbonblack_logo.svg",
+    "description": "Gain extensive insight into VMware Carbon Black Cloud - Endpoint Standard by analyzing, collecting and correlating Event logs.\nThis workbook provides visibility into Carbon Black managed endpoints and identified threat event",
+    "dataTypesDependencies": [ "CarbonBlackEvents_CL" ],
+    "dataConnectorsDependencies": [ "VMware Carbon Black" ],
```
Comment (preetikr) [inline Workbooks/WorkbooksMetadata.json:748]: The data connector dependencies name cannot have spaces in the name- this needs to be equal to the data connector ID value you have in the data connector json file. Please fix this for all instances.

File: Workbooks/WorkbooksMetadata.json [hunk 2]
```diff
@@ -739,5 +739,44 @@
     "templateRelativePath": "SysmonThreatHunting.json",
     "subtitle": "",
     "provider": "Edoardo Gerosa"
+  },
+  {
+    "workbookKey": "VMwareCarbonBlackWorkbook",
+    "logoFileName": "vmwarecarbonblack_logo.svg",
+    "description": "Gain extensive insight into VMware Carbon Black Cloud - Endpoint Standard by analyzing, collecting and correlating Event logs.\nThis workbook provides visibility into Carbon Black managed endpoints and identified threat event",
+    "dataTypesDependencies": [ "CarbonBlackEvents_CL" ],
```
Comment (preetikr) [inline Workbooks/WorkbooksMetadata.json:747]: You can add commas to separate multiple data types - I see Carbon Black has multiple tables, please include all in the dependencies. Same for Proofpoint and others as well - extend this feedback to all.

**Comment by preetikr** [comment]
I've reviewed the data connectors, Azure Functions and related scripts, workbooks and sample data.

@shainw - Please review the analytic rule templates. I see the yaml file custom data type dependencies are inconsistent - it'll be great if you can recommend the right values for requiredDataConnectors:  connectorId: and  dataTypes:
 Thanks!

File: Workbooks/WorkbooksMetadata.json (hunk 1 above)
Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: corrections have been made, updated all data connector json files as well.

File: Workbooks/WorkbooksMetadata.json (hunk 2 above)
Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: corrections have been made

File: DataConnectors/Proofpoint TAP/ProofpointTAP_API_FunctionApp.json [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            "title": "",
+            "description": ">**(Optional Step)** Securely store workspace and API authorization key(s) or token(s) in Azure Key Vault. Azure Key Vault provides a secure mechanism to store and retrieve key values. [Follow these instructions](https://docs.microsoft.com/azure/app-service/app-service-key-vault-references) to use Azure Key Vault with an Azure Function App."
+        },
+        {
+            "title": "",
+            "description": "**STEP 1 - Configuration steps for the Proofpoint TAP API**\n\n1. Log into the Proofpoint TAP console \n2. Navigate to **Connect Applications** and select **Service Principal**\n3. Create a **Service Principal** (API Authorization Key)"
+        },
+        {
+            "title": "",
+            "description": "**STEP 2 - Choose ONE from the following two deployment options to deploy the connector and the associated Azure Function**\n\n>**IMPORTANT:** Before deploying the Proofpoint TAP connector, have the Workspace ID and Workspace Primary Key (can be copied from the following), as well as the Proofpoint TAP API Authorization Key(s), readily available.",
```
[the remark this answers is not in the document]
Comment (chicduong) [reply DataConnectors/Proofpoint TAP/ProofpointTAP_API_FunctionApp.json:107]: per email, these two deployment methods are the simplest to deploy for powershell code. VS Code / VS deployments requires numerous per-requisites to deploy to Azure.

File: DataConnectors/Proofpoint TAP/ProofpointTAP_API_FunctionApp.json [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                "name": "Proofpoint TAP API Key",
+                "description": "A Proofpoint TAP API username and password is required. [See the documentation to learn more about Proofpoint SIEM API](https://help.proofpoint.com/Threat_Insight_Dashboard/API_Documentation/SIEM_API)."
+            }
+        ]
+    },
+    "instructionSteps": [
+        {
+            "title": "",
+            "description": ">**NOTE:** This connector uses Azure Functions to connect to Proofpoint TAP to pull its logs into Azure Sentinel. This might result in additional data ingestion costs. Check the [Azure Functions pricing page](https://azure.microsoft.com/pricing/details/functions/) for details."
+        },
```
[the remark this answers is not in the document]
Comment (chicduong) [reply DataConnectors/Proofpoint TAP/ProofpointTAP_API_FunctionApp.json:96]: per email, these two deployment methods are the simplest to deploy for powershell code. VS Code / VS deployments requires numerous per-requisites to deploy to Azure.

File: Detections/CarbonBlack/KnownMalwareDetected.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+id: 9f86885f-f31f-4e66-a39d-352771ee789e
+name: Known Malware Detected
+description: |
+  'This creates an incident when a known Malware is detected on a endpoint managed by a Carbon Black.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: CarbonBlackEvents_CL
```
Comment (shainw) [inline Detections/CarbonBlack/KnownMalwareDetected.yaml:7]: Same as previous for the connectorId and dataTypes.

File: Detections/ProofpointTAP/MalwareLinkClicked.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+id: 8675dd7a-795e-4d56-a79c-fc848c5ee61c
+name: Malware Link Clicked
+description: |
+  'Creates an incident in the event a user clicks on an email link that is classified as a malware.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: CustomConnector
```
Comment (shainw) [inline Detections/ProofpointTAP/MalwareLinkClicked.yaml:7]: Same as previous for the connectorId and dataTypes

File: Detections/QualysVM/HighNumberofVulnDetected.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+id: be52662c-3b23-435a-a6fa-f39bdfc849e6
+name: High Number of Urgent Vulnerabilities Detected
+description: |
+  'This Creates an incident when a host has a high number of Urgent, severity 5, vulnerabilities detected.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: CustomConnector
```
Comment (shainw) [inline Detections/QualysVM/HighNumberofVulnDetected.yaml:7]: This is not entirely clear and so I will update documentation, but the connectorId maps from the id in your function app - /Qualys VM/QualysVM_API_FunctionApp.json - specifically the first line -

"id": "QualysVulnerabilityManagement",

and later for dataTypes = "name": "QualysHostDetection_CL",

- connectorId: QualysVulnerabilityManagement
dataTypes:
- QualysHostDetection_CL

File: Detections/QualysVM/HighNumberofVulnDetected.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    dataTypes: 
+      - QualysHostDetection_CL
+queryFrequency: 1h
+queryPeriod: 1h
+triggerOperator: gt
+triggerThreshold: 0
+tactics:
```
Comment (shainw) [inline Detections/QualysVM/HighNumberofVulnDetected.yaml:14]: Tactics and Techniques are required - https://attack.mitre.org/techniques/enterprise/

File: Detections/QualysVM/HighNumberofVulnDetected.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+relevantTechniques:
+  - 
+query: |
+
+  let timeframe = ago(1h);
+  QualysHostDetection_CL
+  | where TimeGenerated >= timeframe
+  | mv-expand todynamic(Detections_s)
+  | where Detections_s.Severity == "5"
+  | summarize count() by NetBios_s, IPAddress
```
Comment (shainw) [inline Detections/QualysVM/HighNumberofVulnDetected.yaml:25]: Timestamp is required, so we should bring thru the min and max - summarize StartTime = min(TimeGenerated), EndTime = max(TimeGenerated)

then | extend timestamp = StartTime

File: Detections/QualysVM/HighNumberofVulnDetected.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  | where count_ >= 10
+  | extend HostCustomEntity = NetBios_s, IPCustomEntity = IPAddress
```
Comment (shainw) [inline Detections/QualysVM/HighNumberofVulnDetected.yaml:27]: Add in | extend timestamp = StartTime once made available in the summarize

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+triggerThreshold: 0
+query: |
+
+  let timeframe = ago(1h);
+  QualysHostDetection_CL
+  | where TimeGenerated >= timeframe
+  | mv-expand todynamic(Detections_s)
+  | extend Status = tostring(Detections_s.Status), Vulnerability = tostring(Detections_s.Results), Severity = tostring(Detections_s.Severity)
+  | where Status =~ "New" and Severity == "5"
+  | summarize dcount(NetBios_s) by tostring(Detections_s.QID)
```
Comment (shainw) [inline Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml:22]: Timestamp is required, so we should bring thru the min and max - summarize StartTime = min(TimeGenerated), EndTime = max(TimeGenerated)

then | extend timestamp = StartTime

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  | where dcount_NetBios_s >= 10
```
Comment (shainw) [inline Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml:23]: Add in | extend timestamp = StartTime once made available in the summarize, if Host is available, we should also bring that thru to support further UX functionality

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
+id: 84cf1d59-f620-4fee-b569-68daf7008b7b
+name: New High Severity Vulnerability Detected Across Multiple Hosts
+description: |
+  'This creates an incident when a new high severity vulnerability is detected across multilple hosts'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: CustomConnector
```
Comment (shainw) [inline Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml:7]: Same as previous for connectorId and dataTypes.

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    dataTypes: 
+      - QualysHostDetection_CL
+queryFrequency: 1h
+queryPeriod: 1h
+triggerOperator: gt
+triggerThreshold: 0
+query: |
```
Comment (shainw) [inline Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml:14]: Tactics and Techniques are required - https://attack.mitre.org/techniques/enterprise/

File: Detections/CarbonBlack/CriticalThreatDetected.yaml [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  'This creates an incident in the event a critical threat was identified on a Carbon Black managed endpoint.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: CarbonBlackNotifications_CL
+    dataTypes: 
+      - CustomConnector
+queryFrequency: 1h
+queryPeriod: 1h
+triggerOperator: gt
+triggerThreshold: 0
```
[the remark this answers is not in the document]
Comment (chicduong) [reply Detections/CarbonBlack/CriticalThreatDetected.yaml:13]: corrections have been made. additional corrections applied to a previously submitted rule Detections/SymantecProxySG/UserAccessSuspiciousURLCategories.yaml

[the remark this answers is not in the document]
Comment (chicduong) [reply Detections/CarbonBlack/CriticalThreatDetected.yaml:13]: The alert is very broad, capturing and event that is considered a critical threat by the tool and doesn't align with any specific MITRE tactic. How should these be handled?

File: Detections/CarbonBlack/CriticalThreatDetected.yaml [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+query: |
+
+  let timeframe = ago(1h);
+  CarbonBlackNotifications_CL
+  | where TimeGenerated > timeframe
+  | where threatHunterInfo_score_d >= 8
```
[the remark this answers is not in the document]
Comment (chicduong) [reply Detections/CarbonBlack/CriticalThreatDetected.yaml]: corrections have been made and pushed.

File: Detections/QualysVM/HighNumberofVulnDetected.yaml [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+  - 
+query: |
+
+  let timeframe = ago(1h);
+  QualysHostDetection_CL
+  | where TimeGenerated >= timeframe
+  | mv-expand todynamic(Detections_s)
+  | where Detections_s.Severity == "5"
+  | summarize count() by NetBios_s, IPAddress
+  | where count_ >= 10
```
[the remark this answers is not in the document]
Comment (chicduong) [reply Detections/QualysVM/HighNumberofVulnDetected.yaml]: corrections have been made and pushed.

File: Detections/ProofpointTAP/MalwareAttachmentDelivered.yaml [hunk 18]
```diff
@@ -0,0 +1,29 @@
+id: 0558155e-4556-447e-9a22-828f2a7de06b
+name: Malware attachment delivered
+description: |
+  'Creates an incident in the event a message containing a malware attachment was delivered.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: CustomConnector
```
[the remark this answers is not in the document]
Comment (chicduong) [reply Detections/ProofpointTAP/MalwareAttachmentDelivered.yaml]: corrections have been made and pushed.

File: Detections/CarbonBlack/CriticalThreatDetected.yaml [hunk 19]
```diff
@@ -0,0 +1,23 @@
+id: 2ca4e7fc-c61a-49e5-9736-5da8035c47e0
+name: Critical Threat Detected
+description: |
+  'This creates an incident in the event a critical threat was identified on a Carbon Black managed endpoint.'
+severity: Medium
+requiredDataConnectors:
+  - connectorId: CarbonBlackNotifications_CL
```
[the remark this answers is not in the document]
Comment (chicduong) [reply Detections/CarbonBlack/CriticalThreatDetected.yaml]: corrections have been made and pushed.

File: Detections/QualysVM/HighNumberofVulnDetected.yaml (hunk 10 above)
Comment (chicduong) [reply Detections/QualysVM/HighNumberofVulnDetected.yaml]: corrections have been made and pushed.

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml (hunk 11 above)
Comment (chicduong) [reply Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml]: corrections have been made and pushed.

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml (hunk 12 above)
Comment (chicduong) [reply Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml]: corrections have been made and pushed.

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml (hunk 13 above)
Comment (chicduong) [reply Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml]: corrections have been made and pushed.

File: Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml (hunk 14 above)
Comment (chicduong) [reply Detections/QualysVM/NewHighSeverityVulnDetectedAcrossMulitpleHosts.yaml:18]: corrections have been made and pushed. Also, applied this correction to a previously submitted analytics rule -- Detections/SymantecProxySG/UserAccessedSuspiciousURLCategories.yaml

File: Detections/QualysVM/HighNumberofVulnDetected.yaml (hunk 7 above)
Comment (chicduong) [reply Detections/QualysVM/HighNumberofVulnDetected.yaml]: corrections have been made and pushed.

File: Detections/QualysVM/HighNumberofVulnDetected.yaml (hunk 8 above)
Comment (chicduong) [reply Detections/QualysVM/HighNumberofVulnDetected.yaml:14]: corrections have been made and pushed.

File: Detections/QualysVM/HighNumberofVulnDetected.yaml (hunk 9 above)
Comment (chicduong) [reply Detections/QualysVM/HighNumberofVulnDetected.yaml]: corrections have been made and pushed.

File: Detections/ProofpointTAP/MalwareLinkClicked.yaml (hunk 6 above)
Comment (chicduong) [reply Detections/ProofpointTAP/MalwareLinkClicked.yaml]: corrections have been made and pushed.

File: Detections/CarbonBlack/KnownMalwareDetected.yaml (hunk 5 above)
Comment (chicduong) [reply Detections/CarbonBlack/KnownMalwareDetected.yaml]: corrections have been made and pushed.

**Comment by chicduong** [comment]
@shainw I made retroactive changes to the InfobloxNIOS, SymantecProxySG, SymantecVIP, PulseConnectSecure analytics rules to address the changes to
the timestamp value and the threshold parameters

File: Detections/CarbonBlack/CriticalThreatDetected.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (shainw) [reply Detections/CarbonBlack/CriticalThreatDetected.yaml:13]: Discussed this further with the engineering team and we are good to leave the tactic and technique out.

File: Workbooks/WorkbooksMetadata.json [hunk 20]
```diff
@@ -688,6 +688,45 @@
...
+    "title": "Pulse Connect Secure",
+    "templateRelativePath": "PulseConnectSecure.json",
+    "subtitle": "",
+    "provider": "Pulse Secure"
+  },
+  {
+    "workbookKey": "InfobloxNIOSWorkbook",
+    "logoFileName": "infoblox_logo.svg",
+    "description": "Gain insight into Infoblox NIOS by analyzing, collecting and correlating DHCP and DNS data.\nThis workbook provides visibility into DHCP and DNS traffic",
+    "dataTypesDependencies": ["InfobloxNIOS"],
```
Comment (preetikr) [inline Workbooks/WorkbooksMetadata.json:708]: dataTypesDependencies needs to be Syslog

File: Workbooks/WorkbooksMetadata.json [hunk 21]
```diff
@@ -688,6 +688,45 @@
...
+    "title": "Infoblox NIOS",
+    "templateRelativePath": "InfobloxNIOS.json",
+    "subtitle": "",
+    "provider": "Infoblox"
+  },
+  {
+    "workbookKey": "SymantecVIPWorkbook",
+    "logoFileName": "symantecVIP_logo.svg",
+    "description": "Gain insight into Symantec VIP by analyzing, collecting and correlating strong authentication data.\nThis workbook provides visibility into user authentications",
+    "dataTypesDependencies": ["SymantecVIP"],
```
Comment (preetikr) [inline Workbooks/WorkbooksMetadata.json:721]: dataTypesDependencies needs to be Syslog

File: Workbooks/WorkbooksMetadata.json [hunk 22]
```diff
@@ -701,7 +740,47 @@
...
     "provider": "Edoardo Gerosa"
   },
-  {    "workbookKey": "GitHubSecurityWorkbook",
+  {
+    "workbookKey": "VMwareCarbonBlackWorkbook",
+    "logoFileName": "vmwarecarbonblack_logo.svg",
+    "description": "Gain extensive insight into VMware Carbon Black Cloud - Endpoint Standard by analyzing, collecting and correlating Event logs.\nThis workbook provides visibility into Carbon Black managed endpoints and identified threat event",
+    "dataTypesDependencies": [ "CarbonBlackEvents_CL","CarbonBlackNotifications_CL","CarbonBlackAuditLogs_CL" ],
+    "dataConnectorsDependencies": [ "VMwareCarbonBlack" ],
+    "previewImagesFileNames": [ "VMwareCarbonBlackWhite.png", "VMwareCarbonBlackBlack.png" ],
```
Comment (preetikr) [inline Workbooks/WorkbooksMetadata.json:749]: I suspect the term CarbonBlack would conflict with the Black and White png names. Please rename the png files for just this one to VMWareCarbonBlack.png and VMWareCarbonWhite.png and change in metadata file as well. That way there's less logic confusion.

File: Workbooks/WorkbooksMetadata.json [hunk 23]
```diff
@@ -679,7 +679,7 @@
     "workbookKey": "SymantecProxySGWorkbook",
     "logoFileName": "",
     "description": "Gain insight into Symantec ProxySG by analyzing, collecting and correlating proxy data.\nThis workbook provides visibility into ProxySG Access logs",
-    "dataTypesDependencies": ["Syslog"],
+    "dataTypesDependencies": ["SymantecProxySG"],
```
[the remark this answers is not in the document]
Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: corrections have been made and changes pushed.

File: Workbooks/WorkbooksMetadata.json [hunk 24]
```diff
@@ -688,6 +688,45 @@
     "subtitle": "",
     "provider": "Symantec"
   },
+  {
+    "workbookKey": "PulseConnectSecureWorkbook",
+    "logoFileName": "",
+    "description": "Gain insight into Pulse Secure VPN by analyzing, collecting and correlating vulnerability data.\nThis workbook provides visibility into user VPN activities",
+    "dataTypesDependencies": ["PulseConnectSecure"],
```
[the remark this answers is not in the document]
Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: corrections have been made and changes pushed.

File: Workbooks/WorkbooksMetadata.json (hunk 20 above)
Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: corrections have been made and changes pushed.

File: Workbooks/WorkbooksMetadata.json (hunk 21 above)
Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: corrections have been made and changes pushed.

File: Workbooks/WorkbooksMetadata.json (hunk 22 above)
Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: corrections have been made and changes pushed.

File: Workbooks/WorkbooksMetadata.json [hunk 25]
```diff
@@ -701,7 +740,47 @@
...
+    "subtitle": "",
+    "provider": "Proofpoint"
+  },
+  {
+    "workbookKey": "QualysVMWorkbook",
+    "logoFileName": "qualys_logo.svg",
+    "description": "Gain insight into Qualys Vulnerability Management by analyzing, collecting and correlating vulnerability data.\nThis workbook provides visibility into vulnerabilities detected from vulnerability scans",
+    "dataTypesDependencies": ["QualysHostDetection_CL"],
+    "dataConnectorsDependencies": [ "QualysVulnerabilityManagement" ],
+    "previewImagesFileNames": [ "QualysVulnerabilityWhite.png", "QualysVulnerabilityBlack.png" ],
```
Comment (preetikr) [inline Workbooks/WorkbooksMetadata.json:775]: The preview image files for Qualys doesn't match with the file names. please fix - thanks, for fixing other changes requested

Comment (chicduong) [reply Workbooks/WorkbooksMetadata.json]: orrections have been made and changes pushed.

**Pull request merged**